receiver/receivercreator/discovery.go

```diff
@@ -0,0 +1,186 @@
+	annotations map[string]string,
+	env observer.EndpointEnv) (*receiverTemplate, error) {
+	var port uint16
+	var p observer.Port
+	err := mapstructure.Decode(env, &p)
+	if err != nil {
+		return nil, fmt.Errorf("could not extract port event: %v", zap.Any("env", env))
+	}
+	if p.Port == 0 {
+		return nil, fmt.Errorf("could not extract port: %v", zap.Any("env", env))
+	}
+	port = p.Port
+	pod := p.Pod
+
+	if !discoveryMetricsEnabled(annotations, otelMetricsHints, fmt.Sprint(port)) {
+		return nil, nil
+	}
+
+	subreceiverKey := getHintAnnotation(annotations, otelMetricsHints, scraperHint, fmt.Sprint(port))
+	if subreceiverKey == "" {
+		// no scraper hint detected
+		return nil, nil
+	}
+	if _, ok := builder.ignoreReceivers[subreceiverKey]; ok {
+		// scraper is ignored
+		return nil, nil
+	}
+	builder.logger.Debug("handling added hinted receiver", zap.Any("subreceiverKey", subreceiverKey))
+
+	defaultEndpoint := getStringEnv(env, endpointConfigKey)
+	userConfMap := getScraperConfFromAnnotations(annotations, defaultEndpoint, fmt.Sprint(port), builder.logger)
+
+	recTemplate, err := newReceiverTemplate(fmt.Sprintf("%v/%v_%v", subreceiverKey, pod.UID, port), userConfMap)
+	recTemplate.signals = receiverSignals{true, false, false}
+
+	return &recTemplate, err
+}
+
+func getScraperConfFromAnnotations(
+	annotations map[string]string,
+	defaultEndpoint, scopeSuffix string,
+	logger *zap.Logger) userConfigMap {
+	conf := userConfigMap{}
+	conf[endpointConfigKey] = defaultEndpoint
+
+	configStr := getHintAnnotation(annotations, otelMetricsHints, configHint, scopeSuffix)
+	if configStr == "" {
+		return conf
+	}
+	if err := yaml.Unmarshal([]byte(configStr), &conf); err != nil {
+		logger.Debug("could not unmarshal configuration from hint", zap.Error(err))
+	}
+
+	val := conf[endpointConfigKey]
+	confEndpoint, ok := val.(string)
+	if !ok {
+		logger.Debug("could not extract configured enpoint")
+		return userConfigMap{}
+	}
+
+	if !strings.Contains(confEndpoint, defaultEndpoint) {
+		logger.Debug("configured enpoint should include Pod's endpoint")
+		return userConfigMap{}
+	}
+
+	return conf
+}
+
+func getHintAnnotation(annotations map[string]string, hintBase string, hintKey string, suffix string) string {
+	// try to scope the hint more on container level by suffixing
+	// with .<port> in case of Port event or # TODO: .<container_name> in case of Pod Container event
+	containerLevelHint := annotations[fmt.Sprintf("%s.%s/%s", hintBase, suffix, hintKey)]
+	if containerLevelHint != "" {
+		return containerLevelHint
+	}
+
+	// if there is no container level hint defined try to use the Pod level hint
+	podHintKey := fmt.Sprintf("%s/%s", hintBase, hintKey)
+	return annotations[podHintKey]
+}
+
+func discoveryMetricsEnabled(annotations map[string]string, hintBase string, scopeSuffix string) bool {
+	hintVal := getHintAnnotation(annotations, hintBase, discoveryEnabledHint, scopeSuffix)
+	return hintVal == "true" || hintVal == ""
```

Comment:
This seems like undocumented support of `""`. I think we should avoid that. Let's support only explicit `true` values?

Reply:
Hm.. Looks like `getHintAnnotation` actually returns `""` is the annotation isn't found. So `hintVal == ""` makes it a bug AFAIU

Reply:
The idea here is that if the `io.opentelemetry.discovery.metrics/enabled` is not explicitly disabled we default to `true`. This is to reduce the number of annotations users need to specify. Only defining the scraper annotation should be enough without explicitly adding the `io.opentelemetry.discovery.metrics/enabled: true`.

This is mentioned at https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/451da6ca3a41a363af963e9d2b3a2087c4fbfac5/receiver/receivercreator/README.md?plain=1#L469.

Reply:
Doesn't `io.opentelemetry.discovery.metrics/enabled: ` give `nil` so users need to specify `""` values explicitly as `io.opentelemetry.discovery.metrics/enabled: ""`. I disagree that this provides more benefit than confusion. Users need to specify the annotation, which is typically copy/paste. So it's better to be explicit on what needs to be copy/pasted. `getHintAnnotation` returning `""` still seems like a bug if we accept `""` here.

Reply:
All right, there is a case to be made here for having the `io.opentelemetry.discovery.metrics/enabled` as required annotation hint. Thinking of it again I agree it would be more consistent and would also be future proof if and when the new `Discovery()` interface of scrapers will happen.

I'm gonna change it in this way.